### PR TITLE
feat(28491): add QF to declare missing properties

### LINKED ISSUE #28491: Declare all missing members quick fix
In #28298 we added a new message that elaborated on several of the missing properties in a type. We should specialize the fix here.

### PR DESCRIPTION
**Pull request by a-tarasyuk**
Fixes #28491

### CODE REVIEW

File: src/services/codefixes/fixAddMissingMember.ts [hunk 1]
```diff
@@ -405,4 +429,85 @@ namespace ts.codefix {
...
+            return factory.createStringLiteral((type as StringLiteralType).value, /* isSingleQuote */ quotePreference === QuotePreference.Single);
+        }
+        if (type.flags & TypeFlags.BooleanLiteral) {
+            return (type === checker.getFalseType() || type === checker.getFalseType(/*fresh*/ true)) ? factory.createFalse() : factory.createTrue();
+        }
+        if (type.flags & TypeFlags.Null) {
+            return factory.createNull();
+        }
+        if (type.flags & TypeFlags.Union) {
+            return firstDefined((type as UnionType).types, t => tryGetInitializerValueFromType(context, checker, importAdder, quotePreference, t));
```
Comment (DanielRosenwasser) [inline src/services/codefixes/fixAddMissingMember.ts:483]: Won't you get the `undefined` identifier in some cases here?

Comment (a-tarasyuk) [reply src/services/codefixes/fixAddMissingMember.ts]: Thanks

File: tests/cases/fourslash/codeFixAddMissingProperties8.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
+/// <reference path='fourslash.ts' />
+
+////class C {
+////    constructor() {}
+////}
+////interface I {
+////    c: C;
+////}
```
Comment (DanielRosenwasser) [inline tests/cases/fourslash/codeFixAddMissingProperties8.ts:8]:
```suggestion
////class C {
////    constructor() {}
////}
////abstract class AbstractC {
////}
////class ClassWithConstructorParams {
////    constructor(a: string, b: number, c: C) {}
////}
////interface I {
////    c: C;
////    ac: AbstractC;
////    cwcp: ClassWithConstructorParams;
////}
```

File: src/services/codefixes/fixAddMissingMember.ts [hunk 3]
```diff
@@ -405,4 +429,87 @@ namespace ts.codefix {
...
+        }
+        if (type.flags & TypeFlags.BooleanLiteral) {
+            return (type === checker.getFalseType() || type === checker.getFalseType(/*fresh*/ true)) ? factory.createFalse() : factory.createTrue();
+        }
+        if (type.flags & TypeFlags.Null) {
+            return factory.createNull();
+        }
+        if (type.flags & TypeFlags.Union) {
+            const expression = firstDefined((type as UnionType).types, t => tryGetInitializerValueFromType(context, checker, importAdder, quotePreference, t));
+            return expression ?? createUndefined();
```
Comment (DanielRosenwasser) [inline src/services/codefixes/fixAddMissingMember.ts:484]: Sorry, what I meant is that not only will you potentially get `undefined` (which might not be the case anymore anyway), but you might also get the result of `createUndefined()` even if another result might be more appropriate. For example, with `undefined | number`.

However, thinking about it some more, maybe that's okay.

File: tests/cases/fourslash/codeFixAddMissingProperties1.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+////interface Foo {
+////    a: number;
+////    b: string;
+////    c: 1;
+////    d: "d";
+////    e: "e1" | "e2";
+////    f(x: number, y: number): void;
+////    g: (x: number, y: number) => void;
+////    h: number[]
```
Comment (DanielRosenwasser) [inline tests/cases/fourslash/codeFixAddMissingProperties1.ts:11]:
```suggestion
////    h: number[];
////    i: bigint;
////    j: undefined | "special-string";
////}
```

**Review by DanielRosenwasser — CHANGES_REQUESTED** [review]
Just want a bit more test coverage.

File: src/services/codefixes/fixAddMissingMember.ts (hunk 3 above)
Comment (a-tarasyuk) [reply src/services/codefixes/fixAddMissingMember.ts:484]: For the similar case `j: undefined | "special-string"` in `codeFixAddMissingProperties1` we will get `"special-string"`

Comment (DanielRosenwasser) [reply src/services/codefixes/fixAddMissingMember.ts:484]: Yeah, that's the case I had in mind. Both are fine, we should just be intentional in the behavior and ensure it's tested.

**Review by DanielRosenwasser — APPROVED** [review]

File: src/services/codefixes/fixAddMissingMember.ts [hunk 5]
```diff
@@ -405,4 +429,87 @@ namespace ts.codefix {
...
+        if (getObjectFlags(type) & ObjectFlags.Class) {
+            const classDeclaration = getClassLikeDeclarationOfSymbol(type.symbol);
+            if (classDeclaration === undefined || hasAbstractModifier(classDeclaration)) return createUndefined();
+
+            const constructorDeclaration = getFirstConstructorWithBody(classDeclaration);
+            if (constructorDeclaration && length(constructorDeclaration.parameters)) return createUndefined();
+
+            return factory.createNewExpression(factory.createIdentifier(type.symbol.name), /*typeArguments*/ undefined, /*argumentsArray*/ undefined);
+        }
+        return createUndefined();
```
Comment (andrewbranch) [inline src/services/codefixes/fixAddMissingMember.ts:509]: Need a case for `TypeFlags.TemplateLiteral`?

Comment (a-tarasyuk) [reply src/services/codefixes/fixAddMissingMember.ts:509]: @andrewbranch Thanks for the feedback. Could you help with an example to cover the case with the `TemplateLiteral`?

**Comment by andrewbranch** [comment]
@typescript-bot pack this

**Comment by andrewbranch** [comment]
No playground link @typescript-bot? ☹️

File: src/services/codefixes/fixAddMissingMember.ts (hunk 5 above)
Comment (andrewbranch) [reply src/services/codefixes/fixAddMissingMember.ts:514]:
```ts
interface I1 {
   cssVariable: `--${string}`;
}

const a: I1 = {};
```

I would probably just throw that into the `string` case and seed it with an empty string. No need to try to come up with a placeholder that actually fits the template 😄

Comment (a-tarasyuk) [reply src/services/codefixes/fixAddMissingMember.ts:514]: @andrewbranch Oke. Thanks. I've added a case for `TemplateLiteral ` and a test that covers the strings quote option.

**Review by andrewbranch — APPROVED** [review]
Thanks!

**Pull request merged**